firepit/firepit.js

```diff
@@ -0,0 +1,568 @@
+      ...process.argv.slice(0, 2),
+      "is:npm",
+      "install",
+      "-g",
+      "npm",
+      config.firebase_tools_package
+    ];
+    const code = await ImitateNPM();
+    if (code) {
+      console.log("Setup from remote host failed due to npm error.");
+      process.exit(code);
+    }
+  } else {
+    debug("Using embedded cache for quick install...");
+    debug(
+      shell
+        .cp("-R", path.join(__dirname, "vendor/*"), nodeModulesPath)
+        .toString()
+    );
+  }
+
+  debug(
+    shell
+      .ln(
+        "-sf",
+        path.join(
+          nodeModulesPath,
+          "node_modules/firebase-tools/lib/bin/firebase.js"
+        ),
+        path.join(binPath, "firebase")
+      )
+      .toString()
+  );
+
+  if (!FindTool("firebase-tools/lib/bin/firebase").length) {
+    console.warn(`firebase-tools setup failed.`);
+    process.exit(2);
+  }
+
+  process.argv = original_argv;
+}
+
+function ImitateFirebaseTools(binPath) {
+  debug("Detected no special flags, calling firebase-tools");
+  return new Promise(resolve => {
+    const cmd = fork(binPath, process.argv.slice(2), {
+      stdio: "inherit",
+      env: { ...process.env, FIREPIT_VERSION: version }
+    });
+    cmd.on("close", code => {
+      debug(`firebase-tools is done.`);
+      resolve(code);
+    });
+  });
+}
+
+async function createRuntimeBinaries() {
+  const runtimeBins = {
+    /* Linux / OSX */
+    shell: `"${unsafeNodePath}"  ${runtimeBinsPath}/shell.js "$@"`,
+    node: `"${unsafeNodePath}"  ${runtimeBinsPath}/node.js "$@"`,
+    npm: `"${unsafeNodePath}" "${
+      FindTool("npm/bin/npm-cli")[0]
+    }" ${npmArgs.join(" ")} "$@"`,
+
+    /* Windows */
+    "node.bat": `@echo off
+"${unsafeNodePath}"  ${runtimeBinsPath}\\node.js %*`,
+    "shell.bat": `@echo off
+"${unsafeNodePath}"  ${runtimeBinsPath}\\shell.js %*`,
+    "npm.bat": `@echo off
+node "${FindTool("npm/bin/npm-cli")[0]}" ${npmArgs.join(" ")}  %*`,
+
+    /* Runtime scripts */
+    "shell.js": `${appendToPath.toString()}\n${getSafeCrossPlatformPath.toString()}\n(${runtime.Script_ShellJS.toString()})()`,
+    "node.js": `(${runtime.Script_NodeJS.toString()})()`,
+
+    /* Config files */
+    npmrc: `prefix = ${installPath}`
+  };
+
+  try {
+    shell.mkdir("-p", runtimeBinsPath);
+  } catch (err) {
+    debug(err);
+  }
+
+  if (!flags["disable-write"]) {
+    Object.keys(runtimeBins).forEach(filename => {
+      const runtimeBinPath = path.join(runtimeBinsPath, filename);
+      try {
+        fs.unlinkSync(runtimeBinPath);
+      } catch (err) {
+        debug(err);
+      }
+      fs.writeFileSync(runtimeBinPath, runtimeBins[filename]);
+      shell.chmod("+x", runtimeBinPath);
+    });
+  }
+  debug("Runtime binaries created.");
+}
+
+/*
+-------------------------------------
+Shared Firepit / Runtime Functions
+
+Are invoked in both Firepit and in the Runtime scripts.
+-------------------------------------
+ */
+
+async function getSafeCrossPlatformPath(isWin, path) {
```

Comment:
We pass in `isWin` every time but that's a global right?  Couldn't we just use the global directly?

Reply:
Nah, if you really want to hate yourself and all software and me check out [`L485`](https://github.com/firebase/firebase-tools/pull/1579/files#diff-4d05c05398fa24d4cebdf1fab3b5e7baR485) where this function has `.toString()` called on it. We basically `ctrl-c ctrl-v` this function and plop it into another script where it is used. As a result, it must take `isWin` as an argument since it's called in multiple contexts.